[MLIR] Handle exceptions

erick-xanadu opened the pull request:
**Context:**  Exceptions raised during run time would trigger segmantation faults.

**Description of the Change:** Adds MLIR pass to check for errors returned from asynchronous functions

**Benefits:** No segfaults.

**Possible Drawbacks:** One thing I noticed is that even without the exception handling transformation, it looks like when asynchronous execution happens in parallel pytests, there is a non-zero probability of failure. This gets reduced the less number of parallel tests are run. I will ask the community and see if there is a way to further diagnose this issue.

**Related GitHub Issues:** This PR does not implement the transformation that would propagate the exception handling algorithm up the call graph. This is because inlining guarantees that all calls to quantum operations happen in the "main" method. But this is something that we might consider in the future, or do the runtime approach instead. Things that are left as TODOs:

* Use a stack and a mutex to preserve error messages from multiple threads
*  Before this PR, the runtime did not raised errors when `qml.CNOT(wires=[x, y])` given `x == y`. There might be other places where we need to raise these exceptions.
 * De-allocate device and quantum register when an exception happens.
 
[sc-52638]

rauletorresc commented:
I see some opportunities to improve this PR:

1. Squash all documentation commits into one, as having them separate does not bring any benefit, instead it would make the commit history longer.
2. I think the style changes should either be squashed into the previous commit where the style was broken, or all the style changes should be squashed into one.
3. Most commits have a one or two words description, this will make it difficult to track the intent of the commit in the future. Isn't this a sign that some commits can be merge into one, which can have a meaningful description?
4. Shouldn't we provide some kind of label to differentiate commits making changes to the code logic from commits that don't do it, like documentation or style commits?
5. I noticed some commits with the description "documentation" have actually made small changes to the code logic. I think those two actions should be split into two different commits.

rauletorresc commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -161,36 +161,23 @@ LogicalResult DetectCallsInAsyncRegionsTransform::matchAndRewrite(LLVM::CallOp c
+    bool validCandidate = isScheduledForTransformation(callOp);
+    return validCandidate ? success() : failure();
+}
+
```
The "match" method changed its implementation. I would suggest moving this change to a different commit, because the commit description says it is related to "Documentation".

erick-xanadu replied:
Yes, due to the previous pass, which annotates:

```
        candidate &&
        hasQnodeAttribute(candidate.value()) 
 ```
 
 with `isScheduledForTransformation` the previous two checks are unnecessary.
 
 I can split the commit if you want.

rauletorresc replied:
No need now. I just think separating the changes according to their nature would ease our review process.

rauletorresc commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -284,29 +279,35 @@ void DetectQnodeTransform::rewrite(LLVM::CallOp callOp, PatternRewriter &rewrite
         PatternRewriter::InsertionGuard insertGuard(rewriter);
         rewriter.setInsertionPointToEnd(failBlock);
         rewriter.create<LLVM::UnreachableOp>(callOp->getLoc());
-        return;
```
The deletion of this "return" should be placed in another commit, as it is not related to "Documentation".

rauletorresc commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -284,29 +279,35 @@ void DetectQnodeTransform::rewrite(LLVM::CallOp callOp, PatternRewriter &rewrite
+    // We will now annotate caller for the next stage.
+    //
+    //     llvm.func caller() attributes { catalyst.preHandleError }
+    scheduleAnalysisForErrorHandling(caller, rewriter);
```
This additional instruction should be placed in a separate commit, as it is not related to "Documentation"

erick-xanadu replied:
I moved this function call from before the if statement to after the if statement. I just wanted the annotation for the next pass to be the last change in this function. This is also why I removed the return statement from the if statement.

rauletorresc replied:
Thanks for clarifying. This clarification could have be a good commit description if it was made in a separate commit. As I am new at the code base, it is difficult for me to understand the motivation even for a minor change.

rauletorresc commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -93,6 +94,17 @@ LLVM::LLVMFuncOp lookupOrCreateAwaitTokenName(ModuleOp);
+                                  SmallVector<LLVM::CallOp> &newCalls, PatternRewriter &rewriter);
+void replaceTerminatorWithUnconditionalJumpToSuccessBlock(SmallVector<Block *> abortBlocks,
+                                                          SmallVector<Block *> successBlocks,
+                                                          PatternRewriter &rewriter);
```
This set of functions were not declared here before. This does not seem to be a "Documentation" change. I would suggest making a different commit for those kind of changes.

erick-xanadu replied:
I am moving functions such that the documentation can read from top to bottom. In order to have the code compile, I just needed to add declaration for functions.

rauletorresc commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -748,6 +748,7 @@ void replaceTerminatorWithUnconditionalJumpToSuccessBlock(SmallVector<Block *> a
         PatternRewriter::InsertionGuard insertGuard(rewriter);
         auto terminator = abort->getTerminator();
         rewriter.setInsertionPoint(terminator);
+        assert(isa<LLVM::UnreachableOp>(terminator));
```
This change affects the logic of the program and must be placed in a different commit that is not documentation

rauletorresc commented on mlir/test/Catalyst/AsyncException.mlir:
```diff
@@ -216,6 +216,6 @@ module {
      // CHECK: llvm.call @__catalyst__host__rt__unrecoverable_error
      llvm.call @abort() : () -> ()
-     llvm.br ^finally
+     llvm.unreachable
```
This change affects the logic of the program and must be placed in a different commit that is not documentation

josh146 commented:
@rauletorresc just for some context, our current strategy for merging PRs is a merge commit, with the commit title/description automatically replaced with the PR title and description :) So commits from a merged branch are not preserved in `main`. 

This also encourages us to have useful and detailed PR descriptions.

rauletorresc reviewed (changes requested):
See my comments here: https://github.com/PennyLaneAI/catalyst/pull/447#issuecomment-1892405503

rauletorresc commented:
> @rauletorresc just for some context, our current strategy for merging PRs is a merge commit, with the commit title/description automatically replaced with the PR title and description :) So commits from a merged branch are not preserved in `main`.
> 
> This also encourages us to have useful and detailed PR descriptions.

Thanks for the clarification! Still, a cleaner commit history would ease the code review process at least. 

In the case of a merged PR, if this is too big, it might make it a bit difficult to track a bug introduced in one of the merged small commits. Just a thought.

maliasadi commented on runtime/lib/capi/RuntimeCAPI.cpp:
```diff
@@ -433,6 +435,8 @@ void __quantum__qis__Rot(double phi, double theta, double omega, QUBIT *qubit, b
 
 void __quantum__qis__CNOT(QUBIT *control, QUBIT *target, bool adjoint)
 {
+    RT_FAIL_IF(control == target,
```
Thanks! Would you be able to add a unit test for this check in `Test_LightningCoreQIS.cpp` for the coverage report?

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/447/commits/f156f31b67feb2db2090b4c3b8d8e3a0cb329e8e

maliasadi commented on runtime/lib/capi/RuntimeCAPI.cpp:
```diff
@@ -81,6 +81,8 @@ void deactivateDevice()
 
 extern "C" {
 
+void __catalyst__host__rt__unrecoverable_error() { RT_FAIL("Unrecoverable error."); }
```
Would you add a simple test for this new method in `Test_LightningCoreQIS.cpp` for the codecov report?

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/447/commits/372f237f8f9c38e548794aaf40bce1d286c86817

maliasadi commented on frontend/test/async_tests/test_async.py:
```diff
@@ -0,0 +1,373 @@
+
+
+def test_exception_in_loop2(backend):
+    "Test exception happening in a loop while one qnode succeeds."
```
:raised_hands:

maliasadi commented on frontend/test/async_tests/test_async.py:
```diff
@@ -0,0 +1,373 @@
+    @qjit(async_qnodes=True)
+    @qml.qnode(qml.device(backend, wires=3))
+    def circuit(n):
+        @while_loop(lambda v: v[0] < v[1])
```
would you think it's necessary to add a test for `catalyst.for_loop`?

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/447/commits/a2fc25d46deb8d40b417497157a96d69feccb01f

maliasadi commented on frontend/test/async_tests/test_async.py:
```diff
@@ -0,0 +1,373 @@
+    # Exception in beginning
+    with pytest.raises(RuntimeError, match=msg):
+        wrapper(0)
+
```
can we add checks within async execution of multiple qnodes in calls to `catalyst.adjoint` and `catalyst.ctrl`?

erick-xanadu replied:
I'll add it, but it shouldn't matter which operation is in the program. The exception handling code happens in the caller of the qnode.

erick-xanadu replied:
Here's adjoint: https://github.com/PennyLaneAI/catalyst/pull/447/commits/402a4555c3f6cb30e7cbcdf62b711468a1048d43

maliasadi commented on frontend/test/async_tests/test_async.py:
```diff
@@ -0,0 +1,373 @@
+    with pytest.raises(RuntimeError, match=msg):
+        wrapper(0)
+
+
```
Do you aware of any corner cases for this exception handling support?

erick-xanadu replied:
Yes, if we have a more complex call graph, we haven't handle that. E.g.,

If we have the following callgraph

```
                      A
                  /         \
                 B         C
```

And exceptions happened in either B or C, then we would be able to handle it. However, if exceptions happened at different levels in the callgraph. E.g.,

```
                      A
                  /         \
                 B         C
                 |
                 D
```

If an exception happened in D, and C is executing asynchronously, then we would exit without waiting for the values of C. The solution is to propagate this analysis up the stack. And have the `llvm.call` to A to B be a `llvm.invoke`. The other alternative would be to do something more on the runtime.

Since the inliner inlines everything except functions containing quantum operations, I believe we are guaranteed to have only callgraphs of the above type.

But this is only the first step, so we can always work to expand support.

maliasadi commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -0,0 +1,1272 @@
+
+// Constants
+
+static constexpr llvm::StringRef qnodeAttr = "qnode";
```
nice! didn't know that you can use `llvm::StringRef` with `constexpr` when building with C++17. It's probably implemented as `std::string_view` instead of `std::string` :thinking: as you can use `constexpr` with `string_view` in C++17. `constexpr std::string` is a C++20 feature.

maliasadi commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -0,0 +1,1272 @@
+
+bool isAsync(LLVM::LLVMFuncOp funcOp)
+{
+    if (!funcOp->hasAttr(passthroughAttr))
+        return false;
```
consider this suggestion for the code consistency:
 ```suggestion
    if (!funcOp->hasAttr(passthroughAttr)) {
        return false;
    }
```

maliasadi commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -0,0 +1,1272 @@
+
+    auto haystack = funcOp->getAttrOfType<ArrayAttr>(passthroughAttr);
+    auto needle = StringAttr::get(funcOp.getContext(), "presplitcoroutine");
+    for (auto maybeNeedle : haystack) {
+        if (maybeNeedle == needle)
+            return true;
+    }
```
```suggestion
    auto it = std::find(haystack.begin(), haystack.end(), needle);
    return (it != haystack.end()) ? true : false;
```

maliasadi commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -0,0 +1,1272 @@
+        if (maybeNeedle == needle)
+            return true;
+    }
+
+    return false;
```
```suggestion
```

maliasadi commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -0,0 +1,1272 @@
+std::optional<LLVM::LLVMFuncOp> getCalleeSafe(LLVM::CallOp callOp);
+
+// Helper functions for matching function names
+bool isFunctionNamed(LLVM::LLVMFuncOp funcOp, llvm::StringRef expectedName);
+bool isAbort(LLVM::LLVMFuncOp funcOp);
+bool callsAbort(LLVM::CallOp callOp);
+bool callsAbort(Operation *possibleCall);
+bool isMlirAsyncRuntimeCreateValue(LLVM::LLVMFuncOp funcOp);
+bool isMlirAsyncRuntimeCreateToken(LLVM::LLVMFuncOp funcOp);
+bool isMlirAsyncRuntimeIsTokenError(LLVM::LLVMFuncOp funcOp);
+bool isMlirAsyncRuntimeIsValueError(LLVM::LLVMFuncOp funcOp);
+bool callsMlirAsyncRuntimeCreateValue(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeCreateToken(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeCreateToken(Operation *possibleCall);
+bool callsMlirAsyncRuntimeIsTokenError(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeIsValueError(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeIsTokenError(Operation *possibleCall);
+bool callsMlirAsyncRuntimeIsValueError(Operation *possibleCall);
+bool hasAbortInBlock(Block *block);
```
Would you like to consider inlining these helper methods? Would you be able to add them in a header file?

erick-xanadu replied:
I prefer to let the compiler decide what to inline. :)

maliasadi commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -0,0 +1,1272 @@
+void insertErrorCalls(std::vector<Value> tokens, std::vector<Value> values, Block *failBlock,
+                      PatternRewriter &rewriter);
+void insertBranchFromFailToSuccessor(Block *fail, Block *success, PatternRewriter &rewriter);
+
```
please consider moving these helper functions to a different file.

erick-xanadu replied:
Moved them in several PRs.

maliasadi commented on runtime/lib/capi/RuntimeCAPI.cpp:
```diff
@@ -81,6 +81,8 @@ void deactivateDevice()
 
 extern "C" {
 
+void __catalyst__host__rt__unrecoverable_error() { RT_FAIL("Unrecoverable error."); }
```
If this function is currently called only within the context of async qnodes, will you interested in updating the message to better convey the association with the issue? 

Maybe :thinking: 
```suggestion
void __catalyst__host__rt__unrecoverable_error() { RT_FAIL("Unrecoverable error from asynchronous execution of multiple quantum programs."); }
```

maliasadi reviewed (changes requested):
Great work with sufficient amount of code comments :raccoon: I've had a few suggestions to address and then we can merge it :rocket:

maliasadi commented:
Don't forget to update the changelog :) You can also add GH label: `ci:build_wheels` after addressing those comments to check wheels before merge.

erick-xanadu commented on mlir/test/Catalyst/AsyncException.mlir:
```diff
@@ -0,0 +1,221 @@
+     llvm.call @abort() : () -> ()
+     llvm.unreachable
+  }
+}
```
Thanks! I'll try to improve them too. There is a lot of room to make them better. :sweat_smile:

erick-xanadu replied:
I've improved the tests quite a bit. I am resolving this issue. Feel free to reopen :)

erick-xanadu commented on mlir/include/Catalyst/Transforms/Passes.td:
```diff
@@ -91,4 +91,15 @@ def QnodeToAsyncLoweringPass : Pass<"qnode-to-async-lowering"> {
 }
+
+def DetectQnodePass : Pass<"detect-qnode"> {
+    let summary = "Detect qnodes";
```
https://github.com/PennyLaneAI/catalyst/pull/447/commits/44ef755c764d36442494d403125bf210aa8ace0b

erick-xanadu commented on frontend/test/async_tests/test_async.py:
https://github.com/PennyLaneAI/catalyst/issues/459 I'll resolve this comment, but feel free to reopen.

erick-xanadu commented on runtime/tests/Test_LightningCoreQIS.cpp:
```diff
@@ -1884,3 +1884,40 @@ TEST_CASE("Test the main porperty of the adjoint quantum operations", "[CoreQIS]
     }
 }
+
+TEST_CASE("Test that an exception is raised unconditionally", "[CoreQIS]")
+{
+    auto devices = getDevices();
+    auto &[rtd_lib, rtd_name, rtd_kwargs] = devices[0];
+
+    __catalyst__rt__initialize();
+    __catalyst__rt__device_init((int8_t *)rtd_lib.c_str(), (int8_t *)rtd_name.c_str(),
+                                (int8_t *)rtd_kwargs.c_str());
+
+    REQUIRE_THROWS_WITH(__catalyst__host__rt__unrecoverable_error(),
+                        Catch::Contains("Unrecoverable error"));
+
+    __catalyst__rt__device_release();
+    __catalyst__rt__finalize();
+}
```
@maliasadi it is still here. You might be looking at an old version?

erick-xanadu commented on frontend/test/async_tests/test_async.py:
```diff
@@ -0,0 +1,373 @@
+    def wrapper():
+        return circuit(0)
+
+    # TODO: Better error messages.
```
It is planned for the second PR. The idea is to propagate the error message from all async circuits (or can we can discuss and settle on something else).

maliasadi reviewed (approved):
Nice work! :1st_place_medal:

The Python Formatter action is still failing, but everything else looks good.

erick-xanadu commented:
> The Python Formatter action is still failing, but everything else looks good.

Can't do anything about that. It only complains because I moved files and moved them back. We keep this one because otherwise we have the bug:

```
--- /home/runner/work/catalyst/catalyst/frontend/test/conftest.py:before	2024-02-13 01:09:28.106767
+++ /home/runner/work/catalyst/catalyst/frontend/test/conftest.py:after	2024-02-13 01:09:30.576114
@@ -18,8 +18,9 @@
 import pytest
 
 try:
+    import tensorflow as tf
+
     import catalyst
-    import tensorflow as tf
 except (ImportError, ModuleNotFoundError) as e:
     tf_available = False
 else:
```

And codefactor complains about files I never touched.

erick-xanadu commented:
@dime10 or @maliasadi could you merge this? I can't merge it due to the code factor and python formatting.

dime10 commented on frontend/test/async_tests/__init__.py:
I think we can remove this file no?

erick-xanadu replied:
Yes, you are right. Thanks! https://github.com/PennyLaneAI/catalyst/pull/447/commits/c75e207b77ad643dfb1099eb026d12b898f99890
